File: _archive/src_python_mvp/services/search_service.py
from typing import List
from src.models import Listing, Unit
from src.services.supabase_service import SupabaseService

def search_units_for_buyer(listing: Listing, supabase_service: SupabaseService) -> List[Unit]:
    # Placeholder for foundation: queries the units table
    # Buy listings look for SELL intents.
    query = supabase_service.client.table("units").select("*").eq("is_active", True).eq("intent", "SELL").eq("unit_type", listing.unit_type.value)
    
    if listing.location:
        query = query.ilike("location", f"%{listing.location}%")
        
    if listing.price: # representing budget
        query = query.lte("price", listing.price)
        
    response = query.order("created_at", desc=True).limit(5).execute()
    return [Unit(**data) for data in response.data]

def format_search_results(units: List[Unit]) -> str:
    if not units:
        return "مفيش حاجة مطابقة لطلبك دلوقتي، بس سجلنا طلبك وهنبلغك أول ما يظهر حاجة مناسبة."
        
    results = ["لقينا العروض دي اللي ممكن تناسبك:"]
    for i, u in enumerate(units, 1):
        line = f"{i}. {u.unit_type.value} في {u.location or 'مكان غير محدد'}"
        if u.price:
            line += f" بسعر {u.price} جنيه"
        if u.specs.get("area"):
            line += f"، بمساحة {u.specs['area']} متر"
        results.append(line)
        
    return "\n".join(results)

File: backend/src/search/search.service.ts
/**
 * Search service — queries units table for buyer matching.
 * Ported from Python src/services/search_service.py with Prisma.
 */
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { Unit, UnitType } from '@prisma/client';
import { ListingDto, UnitDto } from '../common/types';

@Injectable()
export class SearchService {
  constructor(private readonly prisma: PrismaService) {}

  /**
   * Search units matching a buyer's criteria.
   * Buy listings look for SELL intents.
   */
  async searchUnitsForBuyer(listing: ListingDto): Promise<Unit[]> {
    return this.prisma.unit.findMany({
      where: {
        isActive: true,
        intent: 'SELL',
        unitType: listing.unitType as UnitType | undefined,
        ...(listing.location
          ? { location: { contains: listing.location } }
          : {}),
        ...(listing.price ? { price: { lte: listing.price } } : {}),
      },
      orderBy: { createdAt: 'desc' },
      take: 5,
    });
  }

  /**
   * Format search results as Ammiya text.
   */
  formatSearchResults(units: UnitDto[]): string {
    if (units.length === 0) {
      return 'مفيش حاجة مطابقة لطلبك دلوقتي، بس سجلنا طلبك وهنبلغك أول ما يظهر حاجة مناسبة.';
    }

    const lines: string[] = ['لقينا العروض دي اللي ممكن تناسبك:'];
    units.forEach((u, i) => {
      let line = `${i + 1}. ${u.unitType} في ${u.location ?? 'مكان غير محدد'}`;
      if (u.price !== null && u.price !== undefined) {
        line += ` بسعر ${u.price} جنيه`;
      }
      const specs = (u.specs ?? {}) as Record<string, unknown>;
      if (specs['area']) {
        line += `، بمساحة ${specs['area']} متر`;
      }
      lines.push(line);
    });

    return lines.join('\n');
  }
}
